dali/python/nvidia/dali/auto_aug/core/_augmentation.py

```diff
@@ -0,0 +1,296 @@
+
+    def _get_magnitudes(self, num_magnitude_bins):
+        mag_range = self.mag_range
+        if mag_range is None:
+            return None
+        if self._has_custom_magnitudes():
+            if num_magnitude_bins is not None and len(mag_range) != num_magnitude_bins:
+                raise Exception(
+                    f"The augmentation `{self.name}` has nd.array of length {len(mag_range)} "
+                    f"specified as the `mag_range`. However, the `num_magnitude_bins` "
+                    f"passed to the call is {num_magnitude_bins}.")
+            return mag_range
+        if num_magnitude_bins is None:
+            raise Exception(
+                f"The `num_magnitude_bins` argument is missing in the call of "
+                f"the `{self.name}` augmentation. Please specify the `num_magnitude_bins` "
+                f"along with the samples and magnitude_bin."
+                f"\nError in augmentation: {self}.")
+        if not hasattr(mag_range, '__len__') or len(mag_range) != 2:
+            raise Exception(
+                f"The `mag_range` must be a tuple of (low, high) ends of magnitude range or "
+                f"nd.array of explicitly defined magnitudes. Got `{self.mag_range}` for "
+                f"augmentation `{self.name}`.")
+        lo, hi = mag_range
+        return np.linspace(lo, hi, num_magnitude_bins, dtype=np.float32)
+
+    def _get_param(self, magnitude_bin, num_magnitude_bins, random_sign=None):
+        assert random_sign is None or isinstance(random_sign, _DataNode)
+        if self.randomly_negate and random_sign is None:
+            warnings.warn(
+                f"The augmentation `{self.name}` was declared with `random_negate=True`, "
+                f"but was called without `random_sign`. A default random source will be "
+                f"created for this call instead. However, if you split batch processing "
+                f"between multiple augmentations, it is best to provide a single common "
+                f"`random_sign` batch manually.", Warning)
+            random_sign = fn.random.uniform(values=[0, 1], dtype=types.INT32)
+        magnitudes = self._get_magnitudes(num_magnitude_bins)
+        if magnitudes is None:
+            return None
+        if magnitude_bin is None:
+            raise Exception(
+                f"The augmentation `{self.name}` has `mag_range` specified, so when called, "
+                f"it requires `magnitude_bin` parameter to select the magnitude from the "
+                f"`mag_range`.\nError in augmentation: {self}.")
+        # if the bin is runtime defined, create Constant node with the array of params
+        # for any possible bin and use subscript to take the right one at each iteration
+        if isinstance(magnitude_bin, _DataNode):
+            if self.randomly_negate:
+                magnitudes = _remap_bins_to_signed_magnitudes(magnitudes)
+                magnitude_bin = 2 * magnitude_bin + random_sign
```

Comment:
Good point, I replaced the explicit random_sign passing with `signed_bin()` call, you invoke it like this `signed_bin(magnitude_bin)` and the necessary node is created at the place of the call. Then, when wrapper sees it as `magnitude_bin` param, it can use that node. It is meant as an optimisation - if user does not make the call but the augmentation has randomly_negate set to True, it will wrap the magnitude_bin on its own.